base/dates/io.jl

```diff
@@ -1,244 +1,219 @@
     end
+    return R(0), i
 end
-function slotformat(slot::Slot{Month},dt,locale)
-    if slot.letter == 'm'
-        return lpad(month(dt),slot.width,"0")
-    elseif slot.letter == 'u'
-        return VALUETOMONTHABBR[locale][month(dt)]
-    else
-        return VALUETOMONTH[locale][month(dt)]
+
+@inline function tryparsenext{N}(d::Delim{String,N}, str, i::Int, len)
+    R = Nullable{Int64}
+    i1=i
+    i2=start(d.d)
+    for j=1:N
+        if i1 > len
+            return R(), i1
+        end
+        c1, i1 = next(str, i1)
+        c2, i2 = next(d.d, i2)
+        if c1 != c2
+            return R(), i1
+        end
     end
+    return R(0), i1
+end
+
+function format(io, d::Delim, dt, locale)
+    write(io, d.d)
 end
-function slotformat(slot::Slot{DayOfWeekSlot},dt,locale)
-    if slot.letter == 'e'
-        return VALUETODAYOFWEEKABBR[locale][dayofweek(dt)]
-    else # == 'E'
-        return VALUETODAYOFWEEK[locale][dayofweek(dt)]
+
+function _show_content{N}(io::IO, d::Delim{Char,N})
+    if d.d in keys(SLOT_RULE)
+        for i=1:N
+            write(io, '\\', d.d)
+        end
+    else
+        for i=1:N
+            write(io, d.d)
+        end
     end
 end
-slotformat(slot::Slot{Millisecond},dt,locale) = rpad(string(millisecond(dt)/1000.0)[3:end], slot.width, "0")
 
-function format(dt::TimeType,df::DateFormat)
-    f = df.prefix
-    for slot in df.slots
-        f *= slotformat(slot,dt,df.locale)
-        if isa(slot, DelimitedSlot)
-            f *= isa(slot.transition, AbstractString) ? slot.transition : ""
+function _show_content(io::IO, d::Delim)
+    for c in d.d
+        if c in keys(SLOT_RULE)
+            write(io, '\\')
         end
+        write(io, c)
     end
-    return f
 end
 
-# UI
-const ISODateTimeFormat = DateFormat("yyyy-mm-dd\\THH:MM:SS.s")
-const ISODateFormat = DateFormat("yyyy-mm-dd")
-const RFC1123Format = DateFormat("e, dd u yyyy HH:MM:SS")
+function Base.show(io::IO, d::Delim)
+    write(io, "Delim(")
+    _show_content(io, d)
+    write(io, ")")
+end
+
+### DateFormat construction
+
+
+abstract DayOfWeekToken # special addition to Period types
+
+# mapping format specifiers to period types
+const SLOT_RULE = Dict{Char, Type}(
+    'y' => Year,
+    'Y' => Year,
+    'm' => Month,
+    'u' => Month,
+    'U' => Month,
+    'e' => DayOfWeekToken,
+    'E' => DayOfWeekToken,
+    'd' => Day,
+    'H' => Hour,
+    'M' => Minute,
+    'S' => Second,
+    's' => Millisecond,
+)
+
+slot_order(::Type{Date}) = (Year, Month, Day)
+slot_order(::Type{DateTime}) = (Year, Month, Day, Hour, Minute, Second, Millisecond)
+
+slot_defaults(::Type{Date}) = map(Int64, (1, 1, 1))
+slot_defaults(::Type{DateTime}) = map(Int64, (1, 1, 1, 0, 0, 0, 0))
+
+slot_types{T<:TimeType}(::Type{T}) = typeof(slot_defaults(T))
 
 """
-    DateTime(dt::AbstractString, format::AbstractString; locale="english") -> DateTime
+    DateFormat(format::AbstractString, locale="english", default_fields=(1,1,1,0,0,0,0)) -> DateFormat
```

Comment:
Should be removed from the docstring. Was needed in an earlier revision of this PR.